pycparser/c_parser.py

```diff
@@ -118,12 +139,16 @@ def __init__(
         # saw: int name;
         # If 'name' is not a key in _scope_stack[n] then 'name' was not defined
         # in this scope at all.
-        self._scope_stack = [dict()]
+
+        # typecheck input and deepcopy twice so we can reset to original
+        # identifiers in parse()
+        self._identifiers = dict([(str(key), bool(val)) for key, val in identifiers.items()])
```

Comment:
I expect that some user will want to feed in their own created dict, so I think it's better to "fail fast" when they supply invalid data, rather than raise a cryptic exception somewhere later in the middle of parsing. It needs to be copied somewhere anyway to have the "child" parser not affect a "parent" parser, so the str() and bool() seems to me a good way to validate input at small extra cost.